jena-arq/src/main/java/org/apache/jena/riot/out/JsonLDWriter.java

```diff
@@ -113,61 +109,45 @@ private void serialize(Writer writer, DatasetGraph dataset, PrefixMap prefixMap,
-                    // dups.add(x) ;
-                } else if ( o.isBlank() || o.isURI() ) {
-                    // add property as a property (the object is an IRI)
-                    Map<String, Object> x2 = new LinkedHashMap<>() ;
-                    x2.put("@id", p.getURI()) ;
-                    x2.put("@type", "@id") ;
-                    ctx.put(x, x2) ;
-                } else if ( o.isLiteral() ) {
-                    String literalDatatypeURI = o.getLiteralDatatypeURI() ;
-                    if ( literalDatatypeURI != null ) {
-                        // add property as a typed attribute (the object is a
-                        // typed literal)
-                        Map<String, Object> x2 = new LinkedHashMap<>() ;
-                        x2.put("@id", p.getURI()) ;
-                        if (! isLangString(o) && ! isSimpleString(o) ) 
-                            // RDF 1.1 : Skip if rdf:langString or xsd:string.
-                            x2.put("@type", literalDatatypeURI) ; 
-                        ctx.put(x, x2) ;
-                    } else {
-                        // add property as an untyped attribute (the object is
-                        // an untyped literal)
-                        ctx.put(x, p.getURI()) ;
-                    }
+        Consumer<Triple> addToContext = triple -> {
+            Node p = triple.getPredicate();
+            Node o = triple.getObject();
+            String localName = p.getLocalName();
+
+            if (o.isBlank() || o.isURI()) {
+                // add property as a property (the object is an IRI)
+                Map<String, Object> x2 = new LinkedHashMap<>();
+                x2.put("@id", p.getURI());
+                x2.put("@type", "@id");
+                ctx.put(localName, x2);
+            } else if (o.isLiteral()) {
+                String literalDatatypeURI = o.getLiteralDatatypeURI();
+                if (literalDatatypeURI != null) {
+                    // add property as a typed attribute (the object is a typed literal)
+                    Map<String, Object> x2 = new LinkedHashMap<>();
+                    x2.put("@id", p.getURI());
+                    if (!(isLangString(o) || isSimpleString(o)))
+                        // RDF 1.1 : Skip if rdf:langString or xsd:string.
+                        x2.put("@type", literalDatatypeURI);
+                    ctx.put(localName, x2);
+                } else {
+                    // add property as an untyped attribute (the object is an untyped literal)
+                    ctx.put(localName, p.getURI());
                 }
             }
-        } ;
-
-        Iter.iter(graph.find(null, null, null)).apply(x) ;
+        };
+        graph.find(ANY)
+            .filterDrop(t -> {
+                Node pred = t.getPredicate();
+                return pred.equals(RDF.type.asNode()) || ctx.containsKey(pred.getLocalName());
+            })
+            .forEachRemaining(addToContext) ;
     }
```

Comment:
I think it was clearer to have the `if ( p.equals(RDF.type.asNode()) )` in teh Consumer because all the logic is in one place.